src/lazy/encoder/mod.rs

```diff
@@ -0,0 +1,567 @@
+    ) -> IonResult<()> {
+        let expected = Element::read_all(expected)?;
+        let mut buffer = Vec::new();
+        let mut writer = LazyRawTextWriter_1_0::new(&mut buffer);
+        test(&mut writer)?;
+        writer.flush()?;
+        let actual = Element::read_all(buffer)?;
+        assert!(IonData::eq(&expected, &actual));
+        Ok(())
+    }
+
+    #[test]
+    fn write_scalars() -> IonResult<()> {
+        let expected = r#"
+            1
+            false
+            3e0
+            "foo"
+            bar
+            2023-11-09T
+            {{4AEA6g==}}
+            [1, 2, 3]
+        "#;
+        let test = |writer: &mut LazyRawTextWriter_1_0<&mut Vec<u8>>| {
+            writer
+                .write(1)?
+                .write(false)?
+                .write(3f32)?
+                .write("foo")?
+                .write("bar".as_symbol_ref())?
+                .write(Timestamp::with_ymd(2023, 11, 9).build()?)?
+                .write([0xE0u8, 0x01, 0x00, 0xEA])?
+                .write([1, 2, 3])?;
+            Ok(())
+        };
+        writer_test(expected, test)
+    }
+
+    #[test]
+    fn write_annotated_scalars() -> IonResult<()> {
+        let expected = r#"
+            foo::bar::1
+            quux::quuz::gary::false
+            Mercury::Venus::3e0
+            Earth::"foo"
+            Mars::Jupiter::bar
+            Saturn::2023-11-09T
+            Uranus::{{4AEA6g==}}
+        "#;
+        let test = |writer: &mut LazyRawTextWriter_1_0<&mut Vec<u8>>| {
+            writer
+                .write(1.annotate(&["foo", "bar"]))?
+                .write(false.annotate(&["quux", "quuz", "gary"]))?
+                .write(3f32.annotate(&["Mercury", "Venus"]))?
+                .write("foo".annotate(&["Earth"]))?
+                .write("bar".as_symbol_ref().annotate(&["Mars", "Jupiter"]))?
+                .write(
+                    Timestamp::with_ymd(2023, 11, 9)
+                        .build()?
+                        .annotate(&["Saturn"]),
+                )?
+                .write((&[0xE0u8, 0x01, 0x00, 0xEA][..]).annotate(&["Uranus"]))?;
+            Ok(())
+        };
+        writer_test(expected, test)
+    }
+
+    #[test]
+    fn write_list() -> IonResult<()> {
+        let expected = r#"
+            [
+              1,
+              false,
+              3e0,
+              "foo",
+              bar,
+              2023-11-09T,
+              {{4AEA6g==}},
+            ]
+        "#;
+        let test = |writer: &mut LazyRawTextWriter_1_0<&mut Vec<u8>>| {
+            let mut list = writer.list_writer()?;
+            list.write(1)?
+                .write(false)?
+                .write(3f32)?
+                .write("foo")?
+                .write("bar".as_symbol_ref())?
+                .write(Timestamp::with_ymd(2023, 11, 9).build()?)?
+                .write(&[0xE0u8, 0x01, 0x00, 0xEA][..])?;
+            list.end()?;
```

Comment:
Just want to confirm—we can chain `end()` after `write()`, like this, right?
```rust
list.write(1)?
     // ...
    .write(&[0xE0u8, 0x01, 0x00, 0xEA][..])?
    .end()?;
```

Reply:
Sadly, no. `write` returns a `&mut Self` and `end` takes (moves) `self`. The alternative is for `write` to _also_ take `self` and return `self`, so we have to choose between one of these:

**Option 1**
```rust
fn write(&mut self) -> IonResult<&mut Self> {...}

// Writing part of a list in a single statement looks nice...
let mut list = writer.list_writer()?;
list.write(1)?
     .write(2)?
     .write(3)?;

//...but we can't write the whole list in a single statement. `end` is on its own line.
let mut list = writer.list_writer()?;
list.write(1)?
    .write(2)?
    .write(3)?
list.end()?;
```

**Option 2**
```rust
fn write(self) -> IonResult<Self> {...}

// Writing the entire list in a single statement looks nice...
writer.list_writer()?
    .write(1)?
     .write(2)?
     .write(3)?
     .end()?;

//...but writing part of a list in a statement requires an extra `list =` prefix
let mut list = writer.list_writer()?;
list = list.write(1)?
    .write(2)?
    .write(3)?
list.end()?;
```

I went with option 1 partly because option 2 makes writing list values in a `for` loop (presumably common) require the extra `list = `.

Open to other ideas! To see another case in which a crate struggled with this tradeoff, see [the `README` for `derive_builder`](https://docs.rs/derive_builder/latest/derive_builder/#owned-aka-consuming).

Reply:
Thanks for the explanation. I agree that option 1 is better.